internal/librarian/generate.go

```diff
@@ -231,14 +261,146 @@ func (r *generateRunner) runBuildCommand(ctx context.Context, outputDir, library
 		LibraryID: libraryID,
 		RepoDir:   r.repo.Dir,
 	}
-
 	slog.Info("Build requested in the context of refined generation; cleaning and copying code to the local language repo before building.")
-	// TODO(https://github.com/googleapis/librarian/issues/775)
-	if err := os.CopyFS(r.repo.Dir, os.DirFS(outputDir)); err != nil {
+	return r.containerClient.Build(ctx, buildRequest)
+}
+
+// clean removes files and directories from a root directory based on remove and preserve patterns.
+//
+// It first determines the paths to remove by applying the removePatterns and then excluding any paths
+// that match the preservePatterns. It then separates the remaining paths into files and directories and
+// removes them, ensuring that directories are removed last.
+//
+// This logic is ported from owlbot logic: https://github.com/googleapis/repo-automation-bots/blob/12dad68640960290910b660e4325630c9ace494b/packages/owl-bot/src/copy-code.ts#L1027
+func clean(rootDir string, removePatterns, preservePatterns []string) error {
+	finalPathsToRemove, err := deriveFinalPathsToRemove(rootDir, removePatterns, preservePatterns)
+	if err != nil {
 		return err
 	}
 
-	return r.containerClient.Build(ctx, buildRequest)
+	filesToRemove, dirsToRemove := separateFilesAndDirs(rootDir, finalPathsToRemove)
+
+	// Remove files first, then directories.
+	for _, file := range filesToRemove {
+		slog.Info("Removing file", "path", file)
+		if err := os.Remove(filepath.Join(rootDir, file)); err != nil {
+			return err
+		}
+	}
+
+	sortDirsByDepth(dirsToRemove)
+
+	for _, dir := range dirsToRemove {
+		slog.Info("Removing directory", "path", dir)
+		if err := os.Remove(filepath.Join(rootDir, dir)); err != nil {
+			// It's possible the directory is not empty due to preserved files.
+			slog.Warn("failed to remove directory, it may not be empty", "dir", dir, "err", err)
+		}
+	}
+
+	return nil
+}
+
+// sortDirsByDepth sorts directories by depth (descending) to remove children first.
+func sortDirsByDepth(dirs []string) {
+	sort.Slice(dirs, func(i, j int) bool {
+		return strings.Count(dirs[i], string(filepath.Separator)) > strings.Count(dirs[j], string(filepath.Separator))
+	})
+}
+
+func getAllPaths(rootDir string) ([]string, error) {
+	var paths []string
+	err := filepath.WalkDir(rootDir, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
```

Comment:
Could you cover this line with a test?

Reply:
done [54be57f](https://github.com/googleapis/librarian/pull/871/commits/54be57f857c296b94bd922ff52f37efb26ab5bb7)